src/enrichment_agent/tools.py

```diff
@@ -5,70 +5,131 @@
+3. Maintain the original facts and data points
+4. Note any missing schema fields that weren't found in the content"""
 
-async def scrape_website(
-    url: str,
-    *,
+async def perform_web_research(
+    query: str, 
+    *, 
     state: Annotated[State, InjectedState],
-    config: Annotated[RunnableConfig, InjectedToolArg],
+    config: Annotated[RunnableConfig, InjectedToolArg]
 ) -> str:
-    """Scrape and summarize content from a given URL.
-
-    Returns:
-        str: A summary of the scraped content, tailored to the extraction schema.
+    """Execute a multi-step web search and information extraction process.
+
+        This function performs the following steps:
+        1. Generates multiple search queries based on the input query
+        2. Executes concurrent web searches using the Tavily API
+        3. Deduplicates and formats the search results
+        4. Extracts structured information based on the provided schema
+
+        Args:
+            query: The initial search query string
+            state: Injected application state containing the extraction schema
+            config: Runtime configuration for the search process
+
+        Returns:
+            str: Structured notes from the search results that are
+             relevant to the extraction schema in state.extraction_schema
+
+        Note:
+            The function uses concurrent execution for multiple search queries to improve
+            performance and combines results from various sources for comprehensive coverage.
     """
-    async with aiohttp.ClientSession() as session:
-        async with session.get(url) as response:
-            content = await response.text()
+    configuration = Configuration.from_runnable_config(config)
 
+    # Generate search queries
+    raw_model = init_model(config)
+    structured_llm = raw_model.with_structured_output(Queries)
+    
+    # Format system instructions
+    query_instructions = query_writer_instructions.format(initial_query=query)
+
+    # Generate queries  
+    results = structured_llm.invoke([SystemMessage(content=query_instructions)]+[HumanMessage(content=f"Please generate {configuration.num_search_queries} search queries.")])
+
+    # Search client
+    tavily_async_client = AsyncTavilyClient()
+
+    # Web search
+    query_list = [query.search_query for query in results.queries]
+    search_tasks = []
+    for query in query_list:
+            search_tasks.append(
+                tavily_async_client.search(
+                    query,
+                    max_results=configuration.max_search_results,
+                    include_raw_content=True,
+                    topic="general"
+                )
+            )
+
+    # Execute all searches concurrently
+    search_docs = await asyncio.gather(*search_tasks)
+
+    # Deduplicate and format sources
+    source_str = deduplicate_and_format_sources(search_docs, max_tokens_per_source=1000, include_raw_content=True)
```

Comment:
probably not needed for v0, but also would be great to propagate sources for each extracted datapoint

Reply:
perhaps can be just included in the extraction schema...